File: OJ/leet1114/leet1114_Semaphore.py
# leet1114

from typing import *


import threading
from multiprocessing import pool 

printFirst = lambda: print('First')
printSecond = lambda: print('Second')
printThird = lambda: print('Third')

from threading import Lock, Semaphore

class Foo:
    def __init__(self):
        self.P12 = Semaphore(0)
        self.P23 = Semaphore(0)

    def first(self, printFirst: 'Callable[[], None]') -> None:
        printFirst()
        self.P12.release()

    def second(self, printSecond: 'Callable[[], None]') -> None:
        with self.P12:
            printSecond()
            self.P23.release()

    def third(self, printThird: 'Callable[[], None]') -> None:

        with self.P23:
            printThird()


 


foo = Foo()
tfirst = threading.Thread(target=foo.first, args=(printFirst,))
tsecond = threading.Thread(target=foo.second, args=(printSecond, ))
tthird = threading.Thread(target=foo.third, args=(printThird,))
tthird2 = threading.Thread(target=foo.third, args=(printThird,))
tthird.start()
tsecond.start()
tfirst.start()

File: OJ/leet1114/leet1114_Semaphore.cpp
/*
leet1114


*/


#include <iostream>
#include <tuple>
#include <vector>
#include <queue>
#include <stack>
#include <string>
#include <set>
#include <map>
#include <unordered_set>
#include <unordered_map>
#include <algorithm>
#include <numeric>
#include <cassert>
#include <random>
#include <functional>
#include <utility>
#include <thread>
#include <mutex>
#include <condition_variable>
#include <semaphore>
#include <future>
#include <atomic>

// #include "semaphore.h"
// #include "../utils/LeetCpp.utils.hpp"

using namespace std;

void printFirst() { cout << "First\n";  }
void printSecond() {  cout << "Second\n"; }
void printThird() {  cout << "Third\n"; }


class Foo {
public:
    binary_semaphore P1{0}, P2{0};

    Foo() { }

    void first(function<void()> printFirst) {
        
        printFirst();

        P1.release();
    }

    void second(function<void()> printSecond) {
        P1.acquire();

        printSecond();

        P2.release();
    }

    void third(function<void()> printThird) {
        P2.acquire();
        
        printThird();
    }
};


int main(int argc, char const *argv[])
{
    Foo foo;
    thread t1(&Foo::first, &foo, printFirst);
    thread t2(&Foo::second, &foo, printSecond);
    thread t3(&Foo::third, &foo, printThird);
    t3.join();
    t2.join();
    t1.join();
    return 0;
}
